### ISSUE TITLE
Advanced indexing gradient is extremely slow when there are many duplicate indices

### ISSUE THREAD

**Issue by sandeepkumar-skb** [pytorch/pytorch#41162]
## 🐛 Bug

We observed a ~2x slowdown in training performance of a detection network when we did broadcast on a tensor.
This broadcasting operation launches the following kernel in the backward pass which completely eats away performance:
https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Indexing.cu#L25
The index of the broadcast has a lot of duplicates and the broadcast is done on the last dimension.

There is no performance loss when the index doesn't contain duplicates. 
Also, it doesn't matter if the index is sorted or unsorted. The code path which calls `indexing_backward_kernel` function sorts the index before computation.

## To Reproduce

1. Here is a python script to see the difference in performance:
```
import torch
from time import time


input_tensor = torch.randn(3, 224, 224, requires_grad=True, device='cuda')
bad_index = torch.randint(19999, 20000, (1, 224, 512), dtype=torch.long, device='cuda')
index = torch.randint(0, 20000, (1, 224, 512), dtype=torch.long, device='cuda')
x = torch.randint(0, 20, [3, 40000], dtype=torch.float32, device='cuda', requires_grad=True)

##### Good Indexing ######
#warming up cache
for i in range(10):
    y = x[:, index]
torch.cuda.synchronize()

t1 = time()
#broadcasting
y = x[:, index]
z = y.sum()
z.backward()
torch.cuda.synchronize()
t1 = time() - t1


##### Bad Indexing ######
#warming up cache
for i in range(10):
    y = x[:, bad_index]
torch.cuda.synchronize()

t2 = time()
#broadcasting
y = x[:, bad_index]
z = y.sum()
z.backward()
torch.cuda.synchronize()
t2 = time() - t2


print("Bad Indexing time: {:.3f}ms".format(t2*1000))
print("Good Indexing time: {:.3f}ms".format(t1*1000))
print("Ratio of Bad Indexing vs Good Indexing: {:.4f}x".format(t2/t1))

```
**Output:**
```
> python indexing_perf_bug.py
Bad Indexing time: 41.887ms
Good Indexing time: 2.191ms
Ratio of Bad Indexing vs Good Indexing: 19.1172x
```
Clearly with duplicate indexing @performance is ~20x worse than the non-duplicate indexing.

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

You can get the script and run it with:
```
Collecting environment information...
PyTorch version: 1.5.0
Is debug build: No
CUDA used to build PyTorch: 10.2

OS: Ubuntu 18.04.4 LTS
GCC version: (Ubuntu 8.4.0-1ubuntu1~18.04) 8.4.0
CMake version: version 3.17.3

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: TITAN RTX
GPU 1: TITAN RTX

Nvidia driver version: 440.100
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip3] numpy==1.16.6
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               10.2.89              hfd86e86_1
[conda] mkl                       2020.0                      166
[conda] mkl-service               2.3.0            py37he904b0f_0
[conda] mkl_fft                   1.0.15           py37ha843d7b_0
[conda] mkl_random                1.1.0            py37hd6b4f25_0
[conda] numpy                     1.18.1           py37h4f9e942_0
[conda] numpy-base                1.18.1           py37hde5b4d6_1
[conda] numpydoc                  0.9.2                      py_0
[conda] pytorch                   1.5.0           py3.7_cuda10.2.89_cudnn7.6.5_0    pytorch
[conda] torchvision               0.6.0                py37_cu102    pytorch
[conda] torchviz                  0.0.1                    pypi_0    pypi
```
@ngimel @mcarilli @csarofeen @ptrblck

**Comment by colesbury** [comment]
The key details are:
- This is in the gradient computation (backward pass)
- There are lots of duplicate indices. The example has 114688 occurrence of a single index. 

I don't think the broadcasting is relevant to the performance issue.

This slow performance is expected given the current implementation because of two constraints:
1) The grad for `y` needs to be summed to just a single value (or to three values with broadcasting)
2) The summation needs to be in a consistent order to achieve bitwise determinism

We can relax (2) now that @kurtamohler has added a `torch.set_deterministic` function. That should help a lot, but having unique indices will still be faster than having lots of duplicates.

In the meantime, you can workaround this by using `Tensor.index_select`, which does not have a bitwise deterministic backwards:

```
x = torch.randint(0, 20, [3, 40000], dtype=torch.float32, device='cuda', requires_grad=True)
bad_index = torch.randint(19999, 20000, (1, 224, 512), dtype=torch.long, device='cuda')
y = x.index_select(1, bad_index.reshape(-1)).reshape(x.shape[:1] + bad_index.shape)
```

**Comment by sandeepkumar-skb** [comment]
Thanks for the suggestion @colesbury. Tried out this suggestion and this indeed brings back the performance!

**Comment by colesbury** [comment]
I'm reopening the issue. I think we should be able to improve the performance of advanced indexing gradient calculation. (Maybe for PyTorch 1.7?)

**Comment by sandeepkumar-skb** [comment]
Sounds great! Thank you.

**Comment by ngimel** [comment]
@colesbury you mean by using nondeterministic algos if user settings allow it?

**Comment by colesbury** [comment]
@ngimel yes

**Comment by anderspapitto** [comment]
it should be possible to speed this up while maintaining bitwise determinism by implicitly using something like a parallel prefix sum (https://en.wikipedia.org/wiki/Prefix_sum#/media/File:Hillis-Steele_Prefix_Sum.svg). 

E.g. instead of summing 10^6 values directly into a single result, sum each 10^3 chunk into each of 10^3 artificial accumulator variables, then sum those 10^3 in a second pass. The max fan-in of a single node is drastically reduced.

This may give a different result than the all at once sum, but it is *deterministic* across multiple runs.

**Comment by anderspapitto** [comment]
here's a proof of concept of manually applying the transformation - but pytorch could do it automatically. 30x speedup, still deterministic

```
import torch

from time import time

bad_index = torch.randint(1999, 2000, (256, 51200), dtype=torch.long, device='cuda')
index = torch.randint(0, 2000, (256, 51200), dtype=torch.long, device='cuda')
x = torch.randint(0, 20, [4000], dtype=torch.float32, device='cuda', requires_grad=True)

##### Good Indexing ######
#warming up cache
for i in range(3):
    y = x[index]
torch.cuda.synchronize()

t1 = time()
y = x[index]
z = y.sum()
z.backward()
torch.cuda.synchronize()
t1 = time() - t1

y_good_simple = y


##### Bad Indexing ######
#warming up cache
for i in range(3):
    y = x[bad_index]
torch.cuda.synchronize()

t2 = time()
y = x[bad_index]
z = y.sum()
z.backward()
torch.cuda.synchronize()
t2 = time() - t2

y_bad_simple = y

streams = [torch.cuda.Stream() for _ in range(256)]

def fixed(idx):
    y = []
    for i in range(idx.shape[0]):
        with torch.cuda.stream(streams[i % len(streams)]):
            y.append(x[idx[i]])
    y = torch.stack(y)
    return y

##### Fixed Indexing ######
#warming up cache
for i in range(3):
    y = fixed(index)
torch.cuda.synchronize()

t3 = time()
y = fixed(index)
z = y.sum()
z.backward()
torch.cuda.synchronize()
t3 = time() - t3

y_good_fixed = y

##### Fixed Bad Indexing ######
#warming up cache
for i in range(3):
    y = fixed(bad_index)
torch.cuda.synchronize()

t4 = time()
y = fixed(bad_index)
z = y.sum()
z.backward()
torch.cuda.synchronize()
t4 = time() - t4

y_bad_fixed = y

print('shapes', y_good_simple.shape, y_good_fixed.shape)
print('shapes', y_bad_simple.shape, y_bad_fixed.shape)
print('simple == fixed?', torch.equal(y_good_simple, y_good_fixed))
print('simple == fixed?', torch.equal(y_bad_simple, y_bad_fixed))

print("Good Indexing time: {:.0f}ms".format(t1*1000))
print("Bad Indexing time: {:.0f}ms".format(t2*1000))
print("Fixed Good Indexing time: {:.0f}ms".format(t3*1000))
print("Fixed Bad Indexing time: {:.0f}ms".format(t4*1000))
print("Ratio of Bad Indexing vs Good Indexing: {:.1f}x".format(t2/t1))
print("Ratio of Fixed Bad Indexing vs Good Indexing: {:.1f}x".format(t4/t1))
print()
print("overall speedup: {:.1f}x".format(t2/t4))
```

```
shapes torch.Size([256, 51200]) torch.Size([256, 51200])
shapes torch.Size([256, 51200]) torch.Size([256, 51200])
simple == fixed? True
simple == fixed? True
Good Indexing time: 140ms
Bad Indexing time: 12186ms
Fixed Good Indexing time: 133ms
Fixed Bad Indexing time: 430ms
Ratio of Bad Indexing vs Good Indexing: 87.3x
Ratio of Fixed Bad Indexing vs Good Indexing: 3.1x

overall speedup: 28.3x
```

**Comment by fmellomascarenhas** [comment]
Hi, any updates on this?
